internal/indexer/oci.go

```diff
@@ -0,0 +1,201 @@
+// 	opt := WithHost("my.oci.domain")
+//
+func WithHost(h string) OciIndexerOpt {
+	return func(opts *ociIndexerOpts) {
+		opts.url = h
+	}
+}
+
+// ociIndexer is an OCI-based Indexer
+type ociIndexer struct {
+	reference string
+	resolver  remotes.Resolver
+}
+
+// DefaultIndexName is the name for the OCI artifact with the index
+const DefaultIndexName = "index"
+
+// DefaultIndexTag is the tag for the OCI artifact with the index
+const DefaultIndexTag = "latest"
+
+// NewOciIndexer returns a new OCI-based indexer
+func NewOciIndexer(opts ...OciIndexerOpt) (Indexer, error) {
+	opt := &ociIndexerOpts{}
+	for _, o := range opts {
+		o(opt)
+	}
+
+	u, err := url.Parse(opt.url)
+	if err != nil {
+		return nil, errors.Wrapf(ErrInvalidArgument, "invalid OCI host URL: %+v", err)
+	}
+	resolver := newDockerResolver(u, opt.username, opt.password, opt.insecure)
+
+	ind := &ociIndexer{
+		reference: opt.reference,
+		resolver:  resolver,
+	}
+	if ind.reference == "" {
+		host := strings.Trim(strings.Join([]string{u.Host, u.Path}, "/"), "/")
+		ind.reference = fmt.Sprintf("%s/%s:%s", host, DefaultIndexName, DefaultIndexTag)
+		klog.Infof("Remote index was not provided. Using default location %q", ind.reference)
+	}
+
+	return ind, nil
+}
+
+// VacAssetIndexLayerMediaType is a media type used in VAC to store a JSON containing the index of
+// charts and containers in a repository
+const VacAssetIndexLayerMediaType = "application/vnd.vmware.tac.asset-index.layer.v1.json"
+// VacAssetIndexConfigMediaType is a media type used in VAC for the configuration of the layer above
+const VacAssetIndexConfigMediaType = "application/vnd.vmware.tac.index.config.v1+json"
+
+// DefaultIndexFilename is the default filename used by the library to upload the index
+const DefaultIndexFilename = "asset-index.json"
+
+// Get implements Indexer
+func (ind *ociIndexer) Get(ctx context.Context) (idx *api.Index, e error) {
+	// Allocate folder for temporary downloads
+	dir, err := os.MkdirTemp("", "indexer")
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to create temporary indexer directory")
+	}
+	defer func() {
+		err := os.RemoveAll(dir)
+		if e == nil {
+			e = err
+		}
+	}()
+
+	indexFile, err := ind.downloadIndex(ctx, dir)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to download index")
+	}
+
+	data, err := ioutil.ReadFile(indexFile)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to read index file")
+	}
+
+	// Populate and return index
+	idx = &api.Index{}
+	if err := pbjson.NewDecoder(bytes.NewReader(data), pbjson.AllowUnknownFields(true)).Decode(idx); err != nil {
+		return nil, errors.Wrapf(err, "unable to parse index file")
+	}
+	return idx, nil
+}
+
+func (ind *ociIndexer) downloadIndex(ctx context.Context, rootPath string) (f string, e error) {
+	// Pull index files from remote
+	store := content.NewFileStore(rootPath)
+	defer func() {
+		err := store.Close()
+		// This library is buggy, and we need to check the error string too
```

Comment:
can you add more details about the error? Were we already using this library before?

Reply:
this is the oras library, you can see how the `Close` function is producing empty errors:

```golang
# oras.land/oras-go/pkg/content/file.go:179

// Close frees up resources used by the file store
func (s *FileStore) Close() error {
	var errs []string
	s.tmpFiles.Range(func(name, _ interface{}) bool {
		if err := os.Remove(name.(string)); err != nil {
			errs = append(errs, err.Error())
		}
		return true
	})
	return errors.New(strings.Join(errs, "; "))
}
```

Reply:
I have filled an issue in the oras-go project: https://github.com/oras-project/oras-go/issues/84